Refactor ttl handling

mikkeloscar opened the pull request:
This changes the TTL handling for removing a PDB in case a statefulset or deployment is not getting ready.

Before we checked `lastTransitionTime` on the individual pods. But because of how Kubernetes keeps updating the status in case you have a pod that is continuously crashing/restarting then we ended up in a situation where the pdb would either not be deleted, or it took much longer than the TTL because it would only happen when the `CrashLoopBackOff` time is long enough to let the TTL kick in.

Instead of relying on `lastTransitionTime` this now relies on the `ReadyReplicas`  of the deployment/statefulset status and we set an annotation: `pdb-controller.zalando.org/non-ready-since` indicating when the resource was last observed with all replicas ready.

This change especially optimizes for the situation where you have pods getting ready for a bit and then getting OOMkilled as we have seen several times with Prometheus. In such cases the `lastTransitionTime` of the pods will be flapping between each pod.

mikkeloscar commented:
If you have a deployment which gets ready and runs for `30s` before it dies then it can still take some time for this to kick in because the statefulset can keep getting ready for some seconds which will reset the `non-ready-since` value. Not sure how to handle this better.

aermakov-zalando commented:
Is it possible to store the annotation on the PDB itself, or would that be too annoying to implement?

mikkeloscar commented:
I moved the annotation to the PDB and also refactored the code a bit by introducing a simple interface:

```go
type kubeResource interface {
	Annotations() map[string]string
	TemplateLabels() map[string]string
	Replicas() int32
	StatusReadyReplicas() int32
}
```

This allows removing duplicate code which was in place to work both with `Deployment` and `StatefulSet` types (yay generics).

linki commented:
:+1:

mikkeloscar commented:
:+1:
